src/NodaTime.Test/DurationTest.Construction.cs

```diff
@@ -165,16 +202,47 @@ public void FromSeconds_Double(double seconds, int expectedDays, long expectedNa
         }
 
         [Test]
+        [TestCase(Duration.MinDays * (double) MillisecondsPerDay, Duration.MinDays, 0)]
         [TestCase(MillisecondsPerDay + MillisecondsPerDay / 2, 1, NanosecondsPerDay / 2)]
         [TestCase(1.5, 0, 1500000)]
         [TestCase(-MillisecondsPerDay + 1.5, -1, 1500000)]
+        [TestCase(MillisecondsPerDay * 8123L + MillisecondsPerSecond + 0.5, 8123, NanosecondsPerSecond + NanosecondsPerMillisecond / 2)]
+        [TestCase(0.01, 0, 10_000)]
+        [TestCase(0.1, 0, 100_000)]
+        [TestCase(4.01, 0, 4_010_000)]
+        [TestCase(4.1, 0, 4_100_000)]
+        [TestCase(4 * MillisecondsPerDay + 0.01, 4, NanosecondsPerMillisecond / 100)]
+        [TestCase(4 * MillisecondsPerDay + 0.1, 4, NanosecondsPerMillisecond / 10)]
+        [TestCase(Duration.MaxDays * (double) MillisecondsPerDay + 0.5, Duration.MaxDays, NanosecondsPerMillisecond / 2)]
         public void FromMilliseconds_Double(double milliseconds, int expectedDays, long expectedNanoOfDay)
         {
             var actual = Duration.FromMilliseconds(milliseconds);
             var expected = new Duration(expectedDays, expectedNanoOfDay);
             Assert.AreEqual(expected, actual);
         }
 
+        [Test]
+        [TestCase(Duration.MinDays * (double) TicksPerDay, Duration.MinDays, 0)]
+        [TestCase(0.01, 0, 1)]
+        [TestCase(0.1, 0, 10)]
+        [TestCase(4.01, 0, 401)]
+        [TestCase(4.1, 0, 410)]
+        [TestCase(4 * TicksPerDay + 0.1, 4, NanosecondsPerTick / 10)]
+        [TestCase(4 * TicksPerDay + 0.01, 4, NanosecondsPerTick / 100)]
+        // Rounding down
+        [TestCase(86_399_999_999_9.994d, 0, NanosecondsPerDay - 1)]
+        // Rounding up across a day boundary
+        [TestCase(86_399_999_999_9.995d, 1, 0)]
+        // We need to add a lot of ticks to the start of "max day" just to get within significant digits.
```

Comment:
This seems fluffy. If you want the next-highest double from a given value, then write that? (Probably something like `BitConverter.UInt64BitsToDouble(BitConverter.DoubleToUInt64Bits(value) + 1)`?)

Reply:
Except we can't write that in an attribute value (because it's not a constant expression) and it's entirely possible that by the time we've done arithmetic, that difference would become irrelevant. I can try moving it into a separate test and see whether it actually works though...